Lib/pathlib.py

```diff
@@ -1193,18 +1236,254 @@ def cwd(cls):
         # os.path.abspath('.') == os.getcwd().
         return cls().absolute()
 
+    def expanduser(self):
+        """ Return a new path with expanded ~ and ~user constructs
+        (as returned by os.path.expanduser)
+        """
+        self._unsupported("expanduser")
+
     @classmethod
     def home(cls):
-        """Return a new path pointing to the user's home directory (as
-        returned by os.path.expanduser('~')).
+        """Return a new path pointing to expanduser('~').
         """
         return cls("~").expanduser()
 
+    def readlink(self):
+        """
+        Return the path to which the symbolic link points.
+        """
+        self._unsupported("readlink")
+
+    def resolve(self, strict=False):
+        """
+        Make the path absolute, resolving all symlinks on the way and also
+        normalizing it.
+        """
+        if self._resolving:
+            return self
+        try:
+            path = self.absolute()
+        except UnsupportedOperation:
+            path = self
+
+        def split(path):
+            return path._from_parsed_parts(path.drive, path.root, []), path._tail[::-1]
+
+        link_count = 0
+        stat_cache = {}
+        target_cache = {}
+        path, parts = split(path)
+        while parts:
+            part = parts.pop()
+            if part == '..':
+                if not path._tail:
+                    if path.root:
+                        # Delete '..' segment immediately following root
+                        continue
+                elif path._tail[-1] != '..':
+                    # Delete '..' segment and its predecessor
+                    path = path.parent
+                    continue
+                path = path._make_child_relpath(part)
+            else:
+                lookup_path = path
+                path = path._make_child_relpath(part)
+                path._resolving = True
+                path_str = str(path)
+                try:
+                    st = stat_cache.get(path_str)
+                    if st is None:
+                        st = stat_cache[path_str] = path.stat(follow_symlinks=False)
```

Comment:
I don't think ``path_str`` can ever be ``None``, so potentially this could use try-except? I suppose it depends on if we expect to hit this cache often or not
```suggestion
                    try:
                        st = stat_cache[path_str]
                    except KeyError:
                        st = stat_cache[path_str] = path.stat(follow_symlinks=False)
```

Reply:
The cache only helps with symlink loops or upward-pointing symlinks, which are an unusual case. Cache _misses_ are much more common in practice.